include/osquery/expected.h

```diff
@@ -0,0 +1,164 @@
+//
+//  expected.hpp
+//  gmock
+//
+//  Created by Max Kareta on 5/15/18.
+//
+
+#pragma once
+
+#include <include/osquery/error.h>
+
+/** Utility class that should be used in function that return
+ * either error or value. Expected enforce developer to test for success and
+ * check error if any.
+ *
+ * ExpectedValue:
+ * ExpectedValue<std::string> function() {
+ *   if (test) {
+ *    return "ok";
+ *   } else {
+ *    return std::shared_prt(new Error(error_domain, error_code));
+ *   }
+ * }
+ *
+ * Expected:
+ * Expected<PlatformProcess> function() {
+ *   if (test) {
+ *    return std::unique_prt(new PlatformProcess(pid));
+ *   } else {
+ *    return std::shared_prt(new Error(error_domain, error_code));
+ *   }
+ * }
+ *
+ * auto result = function();
+ * if (result) {
+ *   ...use *result
+ * } else {
+ *   auto error = result->getError();
+ * }
+ */
+
+namespace osquery {
+
+class ExpectedBase {
+ public:
+  ExpectedBase(Error* error) = delete;
+  ExpectedBase(Error error) = delete;
+  ExpectedBase() : error_(nullptr), hasError_(false) {}
+  ExpectedBase(std::shared_ptr<Error> error)
+      : error_(std::move(error)), hasError_(true) {}
+  ExpectedBase(std::unique_ptr<Error> error)
+      : error_(std::move(error)), hasError_(true) {}
+
+  ~ExpectedBase() {
+    assert(errorChecked_ && (!hasError_ || errorUsed_));
+  }
+
+  std::shared_ptr<Error> getError() {
+    errorUsed_ = true;
+    errorChecked_ = true;
+    return error_;
+  }
+
+  explicit operator bool() {
```

Comment:
Do you think it is a good idea to make `bool()` operator non-const? Why do we need it?

Reply:
it's only way to check success and we need to make sure that it was checked.

Reply:
May it better to define `errorChecked_` as `mutable` member and make `bool()` const method?

Reply:
Looks like perfect case to declare `errorChecked_` mutable and make operator const. It is good when something do exactly what it looks like.